daphne/src/testing.rs

```diff
@@ -467,37 +467,104 @@ impl AuditLog for MockAuditLog {
+    /// results in a valid tree traversal, that is, each prefix encoded by the aggregation
+    /// parameter is equal to or a child of a prefix in the current aggregation parameter.
+    ///
+    /// For simple VDAFs (e.g., Prio3), only the empty aggregation parameter is valid.
+    //
+    // TODO heavy hitters: Enforce the maximum tree depth.
+    pub fn is_valid(&self, next_agg_param: &DapAggregationParam) -> bool {
+        if matches!(self.agg_param, DapAggregationParam::Empty) {
+            return matches!(next_agg_param, DapAggregationParam::Empty);
+        }
+
+        let curr_level = self.agg_param.level();
+        let next_level = next_agg_param.level();
+
+        // Ensure the aggregation level increases monotonically.
+        if next_level < curr_level {
+            return false;
+        }
+
+        // If the level is the same, then ensure the parameter hasn't changed.
+        if next_level == curr_level && next_agg_param != &self.agg_param {
+            return false;
+        }
+
+        // If the next level is greater than the previous, then check that the next parameter is
+        // consistent with the current one.
+        if next_level > curr_level {
+            todo!("TODO heavy hitters: prefix consistency check")
+        }
+
+        true
+    }
+
+    /// If the given aggregation parameter is valid, then advance to it.
+    pub fn goto(&mut self, agg_param: &DapAggregationParam) -> Option<&mut Self> {
+        if !self.is_valid(agg_param) {
+            return None;
+        }
+
+        if agg_param.level() > self.agg_param.level() {
+            // Advance to the next level.
+            self.agg_share.reset();
+            self.collected = false;
+            self.reports.clear();
+            self.agg_param = agg_param.clone();
+        }
+
+        Some(self)
+    }
 }
 
 #[derive(Default)]
 #[cfg_attr(any(test, feature = "test-utils"), derive(deepsize::DeepSizeOf))]
-pub(crate) struct InMemoryAggregateStore(HashMap<String, AggregateStoreForCollection>);
+pub(crate) struct InMemoryAggregateStore(HashMap<String, AggregateStore>);
 
 impl InMemoryAggregateStore {
-    /// Return the aggregate store for the given "collection". A collection is the result of
-    /// aggregating a bucket of reports with a particular aggregation parameter.
-    pub(crate) fn for_collection(
+    /// Return the aggregate store for the given bucket.
+    pub(crate) fn for_bucket(
         &mut self,
         task_id: &TaskId,
         bucket: &DapBatchBucket,
         agg_param: &DapAggregationParam,
-    ) -> &mut AggregateStoreForCollection {
-        let agg_level = agg_param.level();
-        // NOTE(cjpatton) `daphne_server::App` will use a similar naming scheme for instances of
-        // the `AggregateStore` DO. However, for backwards compatibility, if `agg_level == 0`, then
-        // we won't include it in the DO name.
-        self.0
-            .entry(format!("{task_id}/{bucket}/{agg_level}"))
```

Comment:
We can't split the buckets by level anymore because we need to know the last agg param to determine if the next one is valid. In fact, we only need one `AggregateStore` for each bucket: it will store the agg share for the current agg param.